### PR TITLE
[X86][CodeGen] Support folding memory broadcast in X86InstrInfo::foldMemoryOperandImpl

### PR DESCRIPTION
**Pull request by KanRobert**

### CODE REVIEW

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 1]
```diff
@@ -8077,6 +8132,80 @@ MachineInstr *X86InstrInfo::foldMemoryOperandImpl(
...
 }
 
+MachineInstr *X86InstrInfo::foldMemoryBroadcast(
+    MachineFunction &MF, MachineInstr &MI, unsigned OpNum,
+    ArrayRef<MachineOperand> MOs, MachineBasicBlock::iterator InsertPt,
+    unsigned BitsSize, Align Alignment, bool AllowCommute) const {
+
+  const X86FoldTableEntry *I = lookupBroadcastFoldTable(MI.getOpcode(), OpNum);
+
+  if (I)
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:8142]: move lookupBroadcastFoldTable call inside if() ?

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 2]
```diff
@@ -8341,7 +8486,10 @@ bool X86InstrInfo::unfoldMemoryOperand(
   unsigned Index = I->Flags & TB_INDEX_MASK;
   bool FoldedLoad = I->Flags & TB_FOLDED_LOAD;
   bool FoldedStore = I->Flags & TB_FOLDED_STORE;
-  bool FoldedBCast = I->Flags & TB_FOLDED_BCAST;
+  unsigned BCastType = I->Flags & TB_FOLDED_BCAST;
+  // FIXME: Support TB_BCAST_SH in getBroadcastOpcode?
+  if (BCastType == TB_BCAST_SH)
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:8491]: Why was this necessary? Should TB_BCAST_W be here as well?

File: llvm/lib/Target/X86/X86InstrFoldTables.h [hunk 3]
```diff
@@ -44,14 +44,20 @@ const X86FoldTableEntry *lookupTwoAddrFoldTable(unsigned RegOp);
...
+const X86FoldTableEntry *lookupBroadcastFoldTable(unsigned RegOp,
+                                                  unsigned OpNum);
+
 // Look up the memory unfolding table entry for this instruction.
 const X86FoldTableEntry *lookupUnfoldTable(unsigned MemOp);
 
 // Look up the broadcast folding table entry for this instruction from
 // the regular memory instruction.
-const X86FoldTableEntry *lookupBroadcastFoldTable(unsigned MemOp,
+const X86FoldTableEntry *lookupBroadcastFoldTableBySize(unsigned MemOp,
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86InstrFoldTables.h:57]: Maybe pull the lookupBroadcastFoldTable/lookupBroadcastFoldTableBySize change out in a separate NFC patch?

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrFoldTables.h:57]: Good suggestion. Done cfb702676cc181877482a282fe7e07109a24dc9d

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp]: Good suggestion. Done.

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:8489]: I added `TB_BCAST_W` entries in `getBroadcastOpcode` in this PR.  We do not have instruction named `VPBROADCASTSH`, and I don't know if we should use `VPBROADCASTW` for `TB_BCAST_SH`. 

If we do not exclude `TB_BCAST_SH` here, the `unreachable` in `getBroadcastOpcode` would be probably hit.

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 4]
```diff
@@ -8377,7 +8523,7 @@ bool X86InstrInfo::unfoldMemoryOperand(
     // memory access is slow above.
 
     unsigned Opc;
-    if (FoldedBCast) {
+    if (BCastType) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:8526]: Won't this fail on TB_BCAST_W?
```c
TB_BCAST_W = 0 << TB_BCAST_TYPE_SHIFT
```

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:8526]: Yes, it will. I am fixing.

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 5]
```diff
@@ -8191,7 +8330,11 @@ bool X86InstrInfo::unfoldMemoryOperand(
   unsigned Index = I->Flags & TB_INDEX_MASK;
   bool FoldedLoad = I->Flags & TB_FOLDED_LOAD;
   bool FoldedStore = I->Flags & TB_FOLDED_STORE;
-  bool FoldedBCast = I->Flags & TB_FOLDED_BCAST;
+  unsigned BCastType = I->Flags & TB_FOLDED_BCAST;
```
[the remark this answers is not in the document]
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp]: Done

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 4 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp]: Done.

**Comment by KanRobert** [comment]
Strange thing... No tests were updated this time. Need to check what happened.

File: llvm/lib/Target/X86/X86InstrFoldTables.cpp [hunk 6]
```diff
@@ -143,6 +143,23 @@ const X86FoldTableEntry *llvm::lookupFoldTable(unsigned RegOp, unsigned OpNum) {
   return lookupFoldTableImpl(FoldTable, RegOp);
 }
 
+const X86FoldTableEntry *
+llvm::lookupBroadcastFoldTable(unsigned RegOp, unsigned OpNum) {
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86InstrFoldTables.cpp:147]: clang-format this

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrFoldTables.cpp]: Done

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by KanRobert** [comment]
Thanks @RKSimon !
